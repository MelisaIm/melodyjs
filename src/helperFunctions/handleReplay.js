import Tone from 'tone';

export default function handleReplay(song, passChord) {
	if (song.melody) {
		const chords = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15];
		const arrayOfChords = [];
		chords.forEach(chord => {
			let chordArray = song.melody.map(array => array[chord]).filter(note => note !== 0);
			arrayOfChords.push(chordArray);
		});
		arrayOfChords.forEach((chord, index) => {
			setTimeout(() => {
				passChord(index);
				poly(chord.length, chord);
			}, 500 * index);
		});
		setTimeout(() => {
			passChord(null);
		}, 500 * 16);
	}
}
function poly(voices, chordArray) {
	let polySynth = new Tone.PolySynth(voices, Tone.Synth).toMaster();
	polySynth.triggerAttackRelease(chordArray, '16n');
}
