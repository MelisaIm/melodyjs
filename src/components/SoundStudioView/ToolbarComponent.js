import React from 'react';
import handleReplay from 'helperFunctions/handleReplay';

export default function ToolbarComponent({
	songId,
	song,
	onReplay,
	onClear,
	onSave,
	onDelete,
	info,
	passChord,
	playingChord,
	history
}) {
	function _handleReplay() {
		handleReplay(song, passChord);
	}

	// fetch current song
	function _handleClear() {
		onClear(songId);
	}

	// send changes to api
	function _handleSave(event) {
		event.preventDefault();
		onSave(song, songId, history);
	}

	// delete song from api
	function _handleDelete() {
		onDelete(songId, history);
	}

	return (
		<div className="ToolbarComponent">
			<h1>Options</h1>
			<button
				className="action-button action-button-replay"
				onClick={_handleReplay}
				disabled={playingChord}>
				{' '}PLAY SONG{' '}
			</button>

			{songId
				? <div className="ToolbarComponent">
						<button
							className="action-button action-button-save"
							onClick={_handleSave}
							disabled={playingChord}>
							{' '}SAVE CHANGES{' '}
						</button>
						<button
							className="action-button action-button-undo"
							onClick={_handleClear}
							disabled={playingChord}>
							{' '}UNDO CHANGES{' '}
						</button>
						<button
							className="action-button action-button-delete"
							onClick={_handleDelete}
							disabled={playingChord}>
							{' '}DELETE SONG{' '}
						</button>
					</div>
				: <div className="ToolbarComponent">
						<button
							className="action-button action-button-save"
							onClick={_handleSave}
							disabled={playingChord}>
							{' '}SAVE NEW{' '}
						</button>
						<button
							className="action-button action-button-clear"
							onClick={_handleClear}
							disabled={playingChord}>
							{' '}CLEAR{' '}
						</button>
					</div>}
		</div>
	);
}
