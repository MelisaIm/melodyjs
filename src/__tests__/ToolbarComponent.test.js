/* eslint-disable import/first */

import React from 'react';
import { shallow, mount } from 'enzyme';
import ToolbarComponent from '../components/SoundStudioView/ToolbarComponent';
import { assert } from 'chai';
jest.mock('helperFunctions/handleReplay');
import handleReplay from 'helperFunctions/handleReplay';

const song = {
	id: 101,
	melody: [
		['C5', 0, 0, 0, 0, 0, 0, 0, 'C5', 0, 0, 0, 0, 0, 0, 0],
		[0, 0, 0, 0, 0, 0, 0, 'D4', 'D4', 'D4', 0, 0, 0, 0, 0, 0],
		[0, 0, 0, 0, 0, 0, 'E4', 0, 'E4', 0, 'E4', 0, 0, 0, 0, 0],
		[0, 0, 0, 0, 0, 'F4', 0, 0, 'F4', 0, 0, 'F4', 0, 0, 0, 0],
		[0, 0, 0, 0, 'G4', 0, 0, 0, 'G4', 0, 0, 0, 'G4', 0, 0, 0],
		[0, 0, 0, 0, 0, 0, 0, 0, 'A4', 0, 0, 0, 0, 0, 0, 0],
		[0, 0, 0, 0, 0, 0, 0, 0, 'B4', 0, 0, 0, 0, 0, 0, 0],
		[0, 0, 0, 0, 0, 0, 0, 0, 'C4', 0, 0, 0, 0, 0, 0, 0],
		[0, 0, 0, 0, 0, 0, 0, 0, 'D3', 0, 0, 0, 0, 0, 0, 0],
		[0, 0, 0, 0, 0, 0, 0, 0, 'E3', 0, 0, 0, 0, 0, 0, 0],
		[0, 0, 0, 0, 0, 0, 0, 0, 'F3', 0, 0, 0, 0, 0, 0, 0],
		[0, 0, 0, 0, 0, 0, 0, 0, 'G3', 0, 0, 0, 0, 0, 0, 0],
		[0, 0, 0, 0, 0, 0, 0, 0, 'A3', 0, 0, 0, 0, 0, 0, 0],
		[0, 0, 0, 0, 0, 0, 0, 0, 'B3', 0, 0, 0, 0, 0, 0, 0],
		[0, 0, 0, 0, 0, 0, 0, 0, 'C3', 0, 0, 0, 0, 0, 0, 0]
	],
	info: {
		title: 'abacus',
		description: 'just some song'
	}
};

const onReplay = jest.fn();
const onClear = jest.fn();
const onSave = jest.fn();
const onDelete = jest.fn();

const shallowWrapper = shallow(
	<ToolbarComponent
		songId={song.id}
		song={song}
		onReplay={onReplay}
		onClear={onClear}
		onSave={onSave}
		onDelete={onDelete}
		data={[]}
	/>
);

const shallowWrapperNoID = shallow(
	<ToolbarComponent
		songId={null}
		song={null}
		onReplay={onReplay}
		onClear={onClear}
		onSave={onSave}
		onDelete={onDelete}
		data={[]}
	/>
);

describe('Testing ToolbarComponent', () => {
	it('should show 4 buttons instead of 3 when there is a song ID', () => {
		const buttons = shallowWrapper.find('.action-button');
		assert.equal(buttons.length, 4);
	});

	it('should call onClear()', () => {
		shallowWrapper.find('.action-button-undo').simulate('click');
		expect(onClear).toHaveBeenCalled();
	});

	it('should call handleReplay()', () => {
		shallowWrapper.find('.action-button-replay').simulate('click');
		expect(handleReplay).toHaveBeenCalled();
	});

	it('should call onSave()', () => {
		shallowWrapper.find('.action-button-save').simulate('click', { preventDefault() {} });
		expect(onSave).toHaveBeenCalled();
	});

	it('should call onDelete()', () => {
		shallowWrapper.find('.action-button-delete').simulate('click');
		expect(onDelete).toHaveBeenCalled();
	});

	it('should show 3 buttons instead of 4 when there is no song ID', () => {
		const buttons = shallowWrapperNoID.find('.action-button');
		assert.equal(buttons.length, 3);
	});

	it('should pass the snapshot test', () => {
		expect(shallowWrapper).toMatchSnapshot();
	});
});
