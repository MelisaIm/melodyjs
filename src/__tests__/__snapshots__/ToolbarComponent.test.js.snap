// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing ToolbarComponent should pass the snapshot test 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="ToolbarComponent"
  >
    <h1>
      Options
    </h1>
    <button
      className="action-button action-button-replay"
      disabled={undefined}
      onClick={[Function]}
    >
       
      PLAY SONG
       
    </button>
    <div
      className="ToolbarComponent"
    >
      <button
        className="action-button action-button-save"
        disabled={undefined}
        onClick={[Function]}
      >
         
        SAVE CHANGES
         
      </button>
      <button
        className="action-button action-button-undo"
        disabled={undefined}
        onClick={[Function]}
      >
         
        UNDO CHANGES
         
      </button>
      <button
        className="action-button action-button-delete"
        disabled={undefined}
        onClick={[Function]}
      >
         
        DELETE SONG
         
      </button>
    </div>
  </div>,
  "nodes": Array [
    <div
      className="ToolbarComponent"
    >
      <h1>
        Options
      </h1>
      <button
        className="action-button action-button-replay"
        disabled={undefined}
        onClick={[Function]}
      >
         
        PLAY SONG
         
      </button>
      <div
        className="ToolbarComponent"
      >
        <button
          className="action-button action-button-save"
          disabled={undefined}
          onClick={[Function]}
        >
           
          SAVE CHANGES
           
        </button>
        <button
          className="action-button action-button-undo"
          disabled={undefined}
          onClick={[Function]}
        >
           
          UNDO CHANGES
           
        </button>
        <button
          className="action-button action-button-delete"
          disabled={undefined}
          onClick={[Function]}
        >
           
          DELETE SONG
           
        </button>
      </div>
    </div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <ToolbarComponent
        data={Array []}
        onClear={[Function]}
        onDelete={[Function]}
        onReplay={[Function]}
        onSave={[Function]}
        song={
          Object {
            "id": 101,
            "info": Object {
              "description": "just some song",
              "title": "abacus",
            },
            "melody": Array [
              Array [
                "C5",
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                "C5",
                0,
                0,
                0,
                0,
                0,
                0,
                0,
              ],
              Array [
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                "D4",
                "D4",
                "D4",
                0,
                0,
                0,
                0,
                0,
                0,
              ],
              Array [
                0,
                0,
                0,
                0,
                0,
                0,
                "E4",
                0,
                "E4",
                0,
                "E4",
                0,
                0,
                0,
                0,
                0,
              ],
              Array [
                0,
                0,
                0,
                0,
                0,
                "F4",
                0,
                0,
                "F4",
                0,
                0,
                "F4",
                0,
                0,
                0,
                0,
              ],
              Array [
                0,
                0,
                0,
                0,
                "G4",
                0,
                0,
                0,
                "G4",
                0,
                0,
                0,
                "G4",
                0,
                0,
                0,
              ],
              Array [
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                "A4",
                0,
                0,
                0,
                0,
                0,
                0,
                0,
              ],
              Array [
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                "B4",
                0,
                0,
                0,
                0,
                0,
                0,
                0,
              ],
              Array [
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                "C4",
                0,
                0,
                0,
                0,
                0,
                0,
                0,
              ],
              Array [
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                "D3",
                0,
                0,
                0,
                0,
                0,
                0,
                0,
              ],
              Array [
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                "E3",
                0,
                0,
                0,
                0,
                0,
                0,
                0,
              ],
              Array [
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                "F3",
                0,
                0,
                0,
                0,
                0,
                0,
                0,
              ],
              Array [
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                "G3",
                0,
                0,
                0,
                0,
                0,
                0,
                0,
              ],
              Array [
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                "A3",
                0,
                0,
                0,
                0,
                0,
                0,
                0,
              ],
              Array [
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                "B3",
                0,
                0,
                0,
                0,
                0,
                0,
                0,
              ],
              Array [
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                "C3",
                0,
                0,
                0,
                0,
                0,
                0,
                0,
              ],
            ],
          }
        }
        songId={101}
      />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "data": Array [],
          "onClear": [Function],
          "onDelete": [Function],
          "onReplay": [Function],
          "onSave": [Function],
          "song": Object {
            "id": 101,
            "info": Object {
              "description": "just some song",
              "title": "abacus",
            },
            "melody": Array [
              Array [
                "C5",
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                "C5",
                0,
                0,
                0,
                0,
                0,
                0,
                0,
              ],
              Array [
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                "D4",
                "D4",
                "D4",
                0,
                0,
                0,
                0,
                0,
                0,
              ],
              Array [
                0,
                0,
                0,
                0,
                0,
                0,
                "E4",
                0,
                "E4",
                0,
                "E4",
                0,
                0,
                0,
                0,
                0,
              ],
              Array [
                0,
                0,
                0,
                0,
                0,
                "F4",
                0,
                0,
                "F4",
                0,
                0,
                "F4",
                0,
                0,
                0,
                0,
              ],
              Array [
                0,
                0,
                0,
                0,
                "G4",
                0,
                0,
                0,
                "G4",
                0,
                0,
                0,
                "G4",
                0,
                0,
                0,
              ],
              Array [
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                "A4",
                0,
                0,
                0,
                0,
                0,
                0,
                0,
              ],
              Array [
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                "B4",
                0,
                0,
                0,
                0,
                0,
                0,
                0,
              ],
              Array [
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                "C4",
                0,
                0,
                0,
                0,
                0,
                0,
                0,
              ],
              Array [
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                "D3",
                0,
                0,
                0,
                0,
                0,
                0,
                0,
              ],
              Array [
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                "E3",
                0,
                0,
                0,
                0,
                0,
                0,
                0,
              ],
              Array [
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                "F3",
                0,
                0,
                0,
                0,
                0,
                0,
                0,
              ],
              Array [
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                "G3",
                0,
                0,
                0,
                0,
                0,
                0,
                0,
              ],
              Array [
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                "A3",
                0,
                0,
                0,
                0,
                0,
                0,
                0,
              ],
              Array [
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                "B3",
                0,
                0,
                0,
                0,
                0,
                0,
                0,
              ],
              Array [
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                "C3",
                0,
                0,
                0,
                0,
                0,
                0,
                0,
              ],
            ],
          },
          "songId": 101,
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className="ToolbarComponent"
        >
          <h1>
            Options
          </h1>
          <button
            className="action-button action-button-replay"
            disabled={undefined}
            onClick={[Function]}
          >
             
            PLAY SONG
             
          </button>
          <div
            className="ToolbarComponent"
          >
            <button
              className="action-button action-button-save"
              disabled={undefined}
              onClick={[Function]}
            >
               
              SAVE CHANGES
               
            </button>
            <button
              className="action-button action-button-undo"
              disabled={undefined}
              onClick={[Function]}
            >
               
              UNDO CHANGES
               
            </button>
            <button
              className="action-button action-button-delete"
              disabled={undefined}
              onClick={[Function]}
            >
               
              DELETE SONG
               
            </button>
          </div>
        </div>,
        "_debugID": 2,
        "_renderedOutput": <div
          className="ToolbarComponent"
        >
          <h1>
            Options
          </h1>
          <button
            className="action-button action-button-replay"
            disabled={undefined}
            onClick={[Function]}
          >
             
            PLAY SONG
             
          </button>
          <div
            className="ToolbarComponent"
          >
            <button
              className="action-button action-button-save"
              disabled={undefined}
              onClick={[Function]}
            >
               
              SAVE CHANGES
               
            </button>
            <button
              className="action-button action-button-undo"
              disabled={undefined}
              onClick={[Function]}
            >
               
              UNDO CHANGES
               
            </button>
            <button
              className="action-button action-button-delete"
              disabled={undefined}
              onClick={[Function]}
            >
               
              DELETE SONG
               
            </button>
          </div>
        </div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <ToolbarComponent
    data={Array []}
    onClear={[Function]}
    onDelete={[Function]}
    onReplay={[Function]}
    onSave={[Function]}
    song={
      Object {
        "id": 101,
        "info": Object {
          "description": "just some song",
          "title": "abacus",
        },
        "melody": Array [
          Array [
            "C5",
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            "C5",
            0,
            0,
            0,
            0,
            0,
            0,
            0,
          ],
          Array [
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            "D4",
            "D4",
            "D4",
            0,
            0,
            0,
            0,
            0,
            0,
          ],
          Array [
            0,
            0,
            0,
            0,
            0,
            0,
            "E4",
            0,
            "E4",
            0,
            "E4",
            0,
            0,
            0,
            0,
            0,
          ],
          Array [
            0,
            0,
            0,
            0,
            0,
            "F4",
            0,
            0,
            "F4",
            0,
            0,
            "F4",
            0,
            0,
            0,
            0,
          ],
          Array [
            0,
            0,
            0,
            0,
            "G4",
            0,
            0,
            0,
            "G4",
            0,
            0,
            0,
            "G4",
            0,
            0,
            0,
          ],
          Array [
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            "A4",
            0,
            0,
            0,
            0,
            0,
            0,
            0,
          ],
          Array [
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            "B4",
            0,
            0,
            0,
            0,
            0,
            0,
            0,
          ],
          Array [
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            "C4",
            0,
            0,
            0,
            0,
            0,
            0,
            0,
          ],
          Array [
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            "D3",
            0,
            0,
            0,
            0,
            0,
            0,
            0,
          ],
          Array [
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            "E3",
            0,
            0,
            0,
            0,
            0,
            0,
            0,
          ],
          Array [
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            "F3",
            0,
            0,
            0,
            0,
            0,
            0,
            0,
          ],
          Array [
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            "G3",
            0,
            0,
            0,
            0,
            0,
            0,
            0,
          ],
          Array [
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            "A3",
            0,
            0,
            0,
            0,
            0,
            0,
            0,
          ],
          Array [
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            "B3",
            0,
            0,
            0,
            0,
            0,
            0,
            0,
          ],
          Array [
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            "C3",
            0,
            0,
            0,
            0,
            0,
            0,
            0,
          ],
        ],
      }
    }
    songId={101}
  />,
}
`;
